control-plane/agents/ha-node/src/detector.rs

```diff
@@ -0,0 +1,193 @@
+use crate::{
+    path_provider::{CachedNvmePathProvider, NvmePathNameCollection},
+    reporter::PathReporter,
+    Cli,
+};
+use futures::future::FutureExt;
+use nvmeadm::nvmf_subsystem::Subsystem;
+use std::{collections::HashMap, sync::Arc};
+use tokio::time::{sleep, Duration};
+
+/// Possible states of every path record.
+#[derive(Debug, Clone)]
+enum PathState {
+    Good,
+    Suspected,
+    Failed,
+}
+
+/// Object that represents a broken/suspected NVMe path.
+#[derive(Debug, Clone)]
+struct PathRecord {
+    /// Used to allow detection of outdated records.
+    epoch: u64,
+    nqn: String,
+    state: PathState,
+    reporter: Arc<PathReporter>,
+}
+
+impl PathRecord {
+    fn new(nqn: String, epoch: u64, reporter: Arc<PathReporter>) -> Self {
+        Self {
+            nqn,
+            epoch,
+            state: PathState::Good,
+            reporter,
+        }
+    }
+
+    fn get_nqn(&self) -> &str {
+        &self.nqn
+    }
+
+    #[inline]
+    fn get_epoch(&self) -> u64 {
+        self.epoch
+    }
+
+    #[inline]
+    fn set_epoch(&mut self, epoch: u64) {
+        self.epoch = epoch;
+    }
+
+    // Trigger state transition based on 'connecting' state of the underlying NVMe controller.
+    fn report_connecting(&mut self) {
+        match self.state {
+            PathState::Good => {
+                self.state = PathState::Suspected;
+                info!("Target {} transitioned into SUSPECTED state", self.nqn);
+            }
+            PathState::Suspected => {
+                self.state = PathState::Failed;
+                error!(
+                    "NVMe path to target {} transitioned into FAILED state",
+                    self.nqn
+                );
+                self.reporter.report_failed_path(self.nqn.clone());
+            }
+            PathState::Failed => {} // Multiple failures don't cause any state transitions.
+        }
+    }
+
+    fn report_live(&mut self) {
+        self.state = PathState::Good;
+    }
+}
+
+#[derive(Debug)]
+pub struct PathFailureDetector {
+    epoch: u64,
+    detection_period: Duration,
+    suspected_paths: HashMap<String, PathRecord>,
+    reporter: Arc<PathReporter>,
```

Comment:
Why Arc? If it is supposed to be shared across thread, shouldn't PathReporter be protected by Mutex?

Reply:
Mutex is not required, as PathReporter does not require mutable references to work properly.
However, we can replace Arc with Rc (since one PathReporter is shared between multiple instances of PathRecord), as we don't need Send semantics since no async context exists.